File: ios/QuerosermB/Presentation/ExchangeList/ExchangeListViewModel.swift
import Foundation

// MARK: - Generic ViewState
enum ViewState<T> {
    case idle
    case loading
    case success(T)
    case empty
    case error(String)

    var isLoading: Bool {
        if case .loading = self { return true }
        return false
    }
}

// MARK: - ExchangeList ViewModel
@MainActor
final class ExchangeListViewModel {
    @Published private(set) var state: ViewState<[Exchange]> = .idle
    @Published private(set) var isLoadingMore = false
    @Published var loadMoreErrorMessage: String?

    private let getExchangeList: GetExchangeListUseCase
    private let pageSize: Int

    private var nextStart = 1
    private var hasMorePages = true

    init(getExchangeList: GetExchangeListUseCase, pageSize: Int = GetExchangeListUseCase.defaultPageSize) {
        self.getExchangeList = getExchangeList
        self.pageSize = pageSize
    }

    /// Evita novo fetch ao reaparecer a lista (ex.: após voltar do detalhe). Pull-to-refresh usa `refresh()`.
    func loadInitialListIfNeeded() async {
        if case .success(let exchanges) = state, !exchanges.isEmpty {
            return
        }
        await loadExchanges()
    }

    func loadExchanges() async {
        guard !state.isLoading else { return }
        resetPaginationForInitialLoad()
        state = .loading
        loadMoreErrorMessage = nil
        await fetchFirstPage()
    }

    func loadMore() async {
        guard hasMorePages, !isLoadingMore else { return }
        if case .loading = state { return }
        guard case .success = state else { return }

        isLoadingMore = true
        loadMoreErrorMessage = nil
        defer { isLoadingMore = false }

        do {
            let page = try await getExchangeList.execute(start: nextStart, limit: pageSize)
            guard case .success(let current) = state else { return }

            hasMorePages = page.hasMore
            nextStart = page.nextStart
            let merged = Self.sortExchangesByVolume(current + page.items)
            state = merged.isEmpty ? .empty : .success(merged)
        } catch let error as NetworkError {
            loadMoreErrorMessage = error.errorDescription ?? Strings.Error.loadMore
        } catch {
            loadMoreErrorMessage = Strings.Error.loadMore
        }
    }

    func refresh() async {
        resetPaginationForInitialLoad()
        loadMoreErrorMessage = nil
        state = .loading
        await fetchFirstPage()
    }

    private func resetPaginationForInitialLoad() {
        nextStart = 1
        hasMorePages = true
    }

    /// Volume 24h vem do `/exchange/info`; ordenação no app substitui `sort=volume_24h` no map.
    /// nil tratado como -1.0 para ficar abaixo de todas as exchanges com volume real (incluindo zero).
    private static func sortExchangesByVolume(_ list: [Exchange]) -> [Exchange] {
        list.sorted { a, b in
            (a.spotVolumeUSD ?? -1.0) > (b.spotVolumeUSD ?? -1.0)
        }
    }

    private func fetchFirstPage() async {
        do {
            let page = try await getExchangeList.execute(start: 1, limit: pageSize)
            hasMorePages = page.hasMore
            nextStart = page.nextStart
            let sorted = Self.sortExchangesByVolume(page.items)
            state = sorted.isEmpty ? .empty : .success(sorted)
        } catch let error as NetworkError {
            state = .error(error.errorDescription ?? Strings.Error.unknown)
        } catch {
            state = .error(Strings.Error.unknown)
        }
    }
}

File: android/app/src/main/java/br/com/querosermb/presentation/exchangelist/ExchangeListViewModel.kt
package br.com.querosermb.presentation.exchangelist

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import br.com.querosermb.domain.model.Exchange
import br.com.querosermb.domain.usecase.GetExchangeListUseCase
import br.com.querosermb.presentation.UiText
import br.com.querosermb.presentation.ViewState
import br.com.querosermb.presentation.utils.toUiText
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class ExchangeListViewModel @Inject constructor(
    private val getExchangeList: GetExchangeListUseCase
) : ViewModel() {

    private val pageSize = GetExchangeListUseCase.DEFAULT_PAGE_SIZE

    private val _state = MutableStateFlow<ViewState<List<Exchange>>>(ViewState.Idle)
    val state: StateFlow<ViewState<List<Exchange>>> = _state.asStateFlow()

    private val _isLoadingMore = MutableStateFlow(false)
    val isLoadingMore: StateFlow<Boolean> = _isLoadingMore.asStateFlow()

    private val _loadMoreError = MutableStateFlow<UiText?>(null)
    val loadMoreError: StateFlow<UiText?> = _loadMoreError.asStateFlow()

    private var nextStart = 1
    private var hasMorePages = true

    fun loadInitialIfNeeded() {
        if (_state.value is ViewState.Success) return
        loadExchanges()
    }

    fun loadExchanges() {
        viewModelScope.launch {
            nextStart = 1
            hasMorePages = true
            _state.value = ViewState.Loading
            fetchPage()
        }
    }

    fun loadMore() {
        if (_isLoadingMore.value) return
        if (_state.value !is ViewState.Success) return
        if (!hasMorePages) return

        viewModelScope.launch {
            _isLoadingMore.value = true
            _loadMoreError.value = null
            try {
                val page = getExchangeList.execute(start = nextStart, limit = pageSize)
                hasMorePages = page.hasMore
                nextStart = page.nextStart
                val current = (_state.value as? ViewState.Success)?.data ?: emptyList()
                val merged = (current + page.items).sortedByDescending { it.spotVolumeUSD ?: -1.0 }
                _state.value = ViewState.Success(merged)
            } catch (e: Exception) {
                _loadMoreError.value = e.toUiText()
            } finally {
                _isLoadingMore.value = false
            }
        }
    }

    fun refresh() {
        loadExchanges()
    }

    private suspend fun fetchPage() {
        try {
            val page = getExchangeList.execute(start = nextStart, limit = pageSize)
            hasMorePages = page.hasMore
            nextStart = page.nextStart
            _state.value = if (page.items.isEmpty()) ViewState.Empty else ViewState.Success(page.items)
        } catch (e: Exception) {
            _state.value = ViewState.Error(e.toUiText())
        }
    }
}
